### PR TITLE
[Documentation] [General Template] Transformers QnA 

### PR DESCRIPTION
**Pull request by dbogunowicz**
- [x] Test all the commands in the doc.
- [x] Add help section

### CODE REVIEW

**Comment by markurtz** [comment]
@dbogunowicz this looks very good, great job on this! Overall I like the layout and the template. We'll need to update wording a bit so added @jeanniefinks on for that to ensure we have a consistent external voice for users

**Comment by dtransposed** [comment]
> @dbogunowicz this looks very good, great job on this! Overall I like the layout and the template. We'll need to update wording a bit so added @jeanniefinks on for that to ensure we have a consistent external voice for users

I am thinking whether we can maybe divide this task into two items:
- agreeing on a common convention across templates (so we can kick off with writing docs for other pipelines asap).
- polishing up the single doc.

**Comment by dbogunowicz** [comment]
> @dbogunowicz this looks very good, great job on this! Overall I like the layout and the template. We'll need to update wording a bit so added @jeanniefinks on for that to ensure we have a consistent external voice for users

I am thinking whether we can maybe divide this task into two items:
- agreeing on a common convention across templates (so we can kick off with writing docs for other pipelines asap).
- polishing up the single doc.

File: src/sparseml/transformers/README_question_answer.md [hunk 1]
```diff
@@ -0,0 +1,162 @@
...
+It is recommended to run Python 3.8 as some of the scripts within the transformers repository require it.
+
+Note: Transformers will not immediately install with this command. Instead, a sparsification-compatible version of Transformers will install on the first invocation of the Transformers code in SparseML.
+
+## Tutorials
+*Pointer to any existing tutorials published on neuralmagic.com.*
+- [Sparsifying BERT Models Using Recipes](https://github.com/neuralmagic/sparseml/blob/main/integrations/huggingface-transformers/tutorials/sparsifying_bert_using_recipes.md)
+- [Sparse Transfer Learning With BERT](https://github.com/neuralmagic/sparseml/blob/main/integrations/huggingface-transformers/tutorials/bert_sparse_transfer_learning.md)
+
+## Getting started
```
Comment (jeanniefinks) [inline src/sparseml/transformers/README_question_answer.md:27]:
```suggestion
## Getting Started
```

File: src/sparseml/transformers/README_question_answer.md [hunk 2]
```diff
@@ -0,0 +1,162 @@
...
+  model_path='./output'
+)
+
+inference = qa_pipeline(question="What's my name?", context="My name is Snorlax")
+
+>> {'score': 0.9947717785835266, 'start': 11, 'end': 18, 'answer': 'Snorlax'}
+```
+
+
+To learn more, please refer to the [appropriate documentation in the DeepSparse repository](https://github.com/neuralmagic/deepsparse/tree/main/examples/huggingface-transformers)
```
Comment (jeanniefinks) [inline src/sparseml/transformers/README_question_answer.md:161]:
```suggestion
To learn more, refer to the [appropriate documentation in the DeepSparse repository](https://github.com/neuralmagic/deepsparse/tree/main/examples/huggingface-transformers)
```

File: src/sparseml/transformers/README_question_answer.md [hunk 3]
```diff
@@ -0,0 +1,162 @@
...
+  --distill_teacher disable \                       # disable knowledge destillation
+  --recipe zoo:nlp/question_answering/bert-base/pytorch/huggingface/squad/pruned-aggressive_98          
+```
+
+### Sparse Transfer Learning
+
+Once you sparsify a model using SparseML, you can easily sparse fine-tune it on a new dataset.
+While you are free to use your backbone, we encourage you to leverage one of our [sparse pre-trained models](https://sparsezoo.neuralmagic.com) to boost your productivity!
+
+In the example below, we fetch a pruned, quantized BERT model, pre-trained on Wikipedia and Bookcorpus datasets. We then fine tune the model to the SQuAD dataset.
```
Comment (jeanniefinks) [inline src/sparseml/transformers/README_question_answer.md:54]:
```suggestion
In the example below, we fetch a pruned, quantized BERT model, pre-trained on Wikipedia and Bookcorpus datasets. We then fine-tune the model to the SQuAD dataset. 
```

File: src/sparseml/transformers/README_question_answer.md [hunk 4]
```diff
@@ -0,0 +1,162 @@
...
+#### Knowledge Distillation
+By modifying the `distill_teacher` argument, you can enable [Knowledge Destillation](https://neptune.ai/blog/knowledge-distillation) (KD) functionality.
+
+In this example, the `--distill_teacher` argument is set to pull a dense SQuAD model from the SparseZoo to enable it to run independently of the dense teacher step:
+
+```bash
+--distill_teacher zoo:nlp/question_answering/bert-base/pytorch/huggingface/squad/base-none
+```
+
+Alternatively, the user may decide to train their own dense teacher model. The following command will use the dense BERT base model from the SparseZoo and fine-tune it on the SQuAD dataset.
```
Comment (jeanniefinks) [inline src/sparseml/transformers/README_question_answer.md:75]:
```suggestion
Alternatively, the user may decide to train their own dense teacher model. The following command uses the dense BERT base model from the SparseZoo and fine-tunes it on the SQuAD dataset.
```

File: src/sparseml/transformers/README_question_answer.md [hunk 5]
```diff
@@ -0,0 +1,162 @@
...
+Once the dense teacher is trained we may reuse it for KD. This only requires passing the path to the directory with the model:
+
+```bash
+--distill_teacher models/teacher
+```
+
+## SparseML CLI
+*This section should explain the functionality of the most important parameters of the CLI. The goal is not to teach the user about every argument, but to make it easy for them to tweak the crucial ones.*
+
+The SparseML installation provides a CLI for sparsifying your models for a specific task; appending the `--help` argument will provide a full list of options for training in SparseML:
```
Comment (jeanniefinks) [inline src/sparseml/transformers/README_question_answer.md:95]:
```suggestion
The SparseML installation provides a CLI for sparsifying your models for a specific task; appending the `--help` argument displays a full list of options for training in SparseML:
```

File: src/sparseml/transformers/README_question_answer.md [hunk 6]
```diff
@@ -0,0 +1,162 @@
...
+
+### DeepSparse Engine Deployment
+
+*No pipeline is complete without its integration with the DeepSparse Engine. This chapter should communicate to the user,
+that they can eventually use their ONNX model to deploy the network using the DeepSparse Engine and benefit from GPU-class performance
+on their CPU! Let's make this section short, referencing internal DeepSparse docs as much as possible.* 
+
+Once the model is exported an ONNX format, it is ready for deployment with the DeepSparse Engine. 
+
+The deployment is intuitive and simple thanks to DeepSparse Python API.
```
Comment (jeanniefinks) [inline src/sparseml/transformers/README_question_answer.md:145]:
```suggestion
The deployment is intuitive due to the DeepSparse Python API.
```

**Review by jeanniefinks — CHANGES_REQUESTED** [review]
For technical writing style, please 

- do not use "please" or "thanks/thank you."
- use present simple tense in most cases

Include how users get help; add as a new paragraph to bottom:

For Neural Magic Support, sign up or log in to our **[Deep Sparse Community Slack](https://join.slack.com/t/discuss-neuralmagic/shared_invite/zt-q1a1cnvo-YBoICSIw3L1dmQpjBeDurQ).** Bugs, feature requests, or additional questions can also be posted to our [GitHub Issue Queue.](https://github.com/neuralmagic/sparseml/issues)

**Comment by dbogunowicz** [comment]
Just a reminder, that we should probably remove my general "template" comments before this lands as a doc for the users.

**Review by bfineran — APPROVED** [review]

**Comment by dbogunowicz** [comment]
@mgoin not really sure who would be a good person to ask, but I guess that's you:
do you have any thoughts on sprinkling some consistent "beautifications" on the docs? I am talking specifically about e.g. adding emoticons to the section headers.

File: src/sparseml/transformers/README_question_answering.md [hunk 7]
```diff
@@ -0,0 +1,165 @@
...
+In the example below, a dense BERT model is trained on the SQuAD dataset. By passing the recipe `zoo:nlp/question_answering/bert-base/pytorch/huggingface/squad/pruned-aggressive_98` (located in [SparseZoo](https://sparsezoo.neuralmagic.com/models/nlp%2Fquestion_answering%2Fbert-base%2Fpytorch%2Fhuggingface%2Fsquad%2Fpruned-aggressive_98)) we modify (sparsify) the training process and/or the model.
+
+```python
+sparseml.transformers.question_answering \
+  --model_name_or_path bert-base-uncased \          # name of the Hugging Face dense model
+  --dataset_name squad \                            # name of the dataset we want to sparse train on
+  --do_train \                                      # run training
+  --do_eval \                                       # run evaluation on validation set 
+  --output_dir './output' \                         # output directory of the saved model
+  --cache_dir cache \                               # local directory to store the downloaded hugging face model."
```
Comment (dbogunowicz) [inline src/sparseml/transformers/README_question_answering.md:44]:
```suggestion
  --cache_dir cache \                               # local directory to store the downloaded hugging face model.  
```

File: src/sparseml/transformers/README_question_answering.md [hunk 8]
```diff
@@ -0,0 +1,165 @@
...
+
+*This subsection should explain to the user how to export their trained, sparse model to ONNX format.*
+
+The DeepSparse Engine uses the ONNX format to load neural networks and then deliver breakthrough performance for CPUs by leveraging the sparsity and quantization within a network.
+
+The SparseML installation provides a `sparseml.transformers.export_onnx` command that you can use to load the training model folder and create a new `model.onnx` file within. Be sure the `--model_path` argument points to your trained model. 
+```bash
+sparseml.transformers.export_onnx \
+    --model_path './output' \
+    --task 'question-answering' \
```
Comment (dbogunowicz) [inline src/sparseml/transformers/README_question_answering.md:134]:
```suggestion
    --task 'question-answering' 
```

**Review by rahul-tuli — DISMISSED** [review]

**Review by KSGulin — COMMENTED** [review]
Thanks for putting together this template @dbogunowicz. It was really helpful in writing the rest of the docs. I left some comments and suggestions.

File: src/sparseml/transformers/README_question_answering.md [hunk 9]
```diff
@@ -0,0 +1,152 @@
+# SparseML Transformer Question Answer Integration
+
+
+SparseML Question Answer pipeline integrates with Hugging Face’s Transformers library to enable the sparsification of any transformer model.
```
[the remark this answers is not in the document]
Comment (dbogunowicz) [reply src/sparseml/transformers/README_question_answering.md:4]: Right.

Comment (dbogunowicz) [inline src/sparseml/transformers/README_question_answering.md:4]:
```suggestion
SparseML Question Answer pipeline integrates with Hugging Face’s Transformers library to enable the sparsification of a large set of transformers models.
```

File: src/sparseml/transformers/README_question_answering.md [hunk 10]
```diff
@@ -0,0 +1,152 @@
...
+## Tutorials
+
+- [Sparsifying BERT Models Using Recipes](https://github.com/neuralmagic/sparseml/blob/main/integrations/huggingface-transformers/tutorials/sparsifying_bert_using_recipes.md)
+- [Sparse Transfer Learning With BERT](https://github.com/neuralmagic/sparseml/blob/main/integrations/huggingface-transformers/tutorials/bert_sparse_transfer_learning.md)
+
+## Getting Started
+
+### Sparsification of Popular Transformer Models
+
+Sparse ML Hugging Face Question Answer integration allows sparsifying any dense transformer.
```
[the remark this answers is not in the document]
Comment (dbogunowicz) [reply src/sparseml/transformers/README_question_answering.md:29]: Good point.

File: src/sparseml/transformers/README_question_answering.md [hunk 11]
```diff
@@ -0,0 +1,152 @@
...
+## Tutorials
+
+- [Sparsifying BERT Models Using Recipes](https://github.com/neuralmagic/sparseml/blob/main/integrations/huggingface-transformers/tutorials/sparsifying_bert_using_recipes.md)
+- [Sparse Transfer Learning With BERT](https://github.com/neuralmagic/sparseml/blob/main/integrations/huggingface-transformers/tutorials/bert_sparse_transfer_learning.md)
+
+## Getting Started
+
+### Sparsifying Popular Transformer Models
+
+Sparse ML Hugging Face Question Answer integration allows sparsifying any dense transformer.
```
Comment (dbogunowicz) [inline src/sparseml/transformers/README_question_answering.md:29]:
```suggestion
```

File: src/sparseml/transformers/README_question_answering.md [hunk 12]
```diff
@@ -0,0 +1,151 @@
+# SparseML Transformer Question Answer Integration
+
+
+SparseML Question Answer pipeline integrates with Hugging Face’s Transformers library to enable the sparsification of a large set of transformers models.
+Sparsification is a powerful technique that results in faster, smaller, and cheaper deployable models. 
+After training, the model can be deployed with Neural Magic's DeepSparse Engine. The engine enables inference with GPU-class performance directly on your CPU.
+
+This integration enables spinning up one of the following end-to-end functionalities:
+- **Sparsification of Popular Transformer Models** - easily sparsify any popular Hugging Face transformer models. 
+- **Sparse Transfer Learning** - fine-tune a sparse backbone model (or use one of our [sparse pre-trained models](https://sparsezoo.neuralmagic.com/?page=1&domain=nlp&sub_domain=question_answering)) on your own, private dataset.
```
Comment (jeanniefinks) [inline src/sparseml/transformers/README_question_answering.md:10]:
```suggestion
- **Sparse Transfer Learning** - fine-tune a sparse backbone model (or use one of our [sparse pre-trained models](https://sparsezoo.neuralmagic.com/?page=1&domain=nlp&sub_domain=question_answering)) on your own private dataset.
```

File: src/sparseml/transformers/README_question_answering.md [hunk 13]
```diff
@@ -0,0 +1,151 @@
+# SparseML Transformer Question Answer Integration
```
Comment (jeanniefinks) [inline src/sparseml/transformers/README_question_answering.md:1]: should it really be:
SparseML Transformer Question Answering Integration

File: src/sparseml/transformers/README_question_answering.md [hunk 14]
```diff
@@ -0,0 +1,151 @@
...
+
+
+SparseML Question Answer pipeline integrates with Hugging Face’s Transformers library to enable the sparsification of a large set of transformers models.
```
Comment (jeanniefinks) [inline src/sparseml/transformers/README_question_answering.md:4]: Question Answer = Question Answering (?)

File: src/sparseml/transformers/README_question_answering.md [hunk 15]
```diff
@@ -0,0 +1,151 @@
...
+    --dataset_name squad \
+    --do_train \
+    --do_eval \
+    --output_dir './output' \ 
+    --distill_teacher disable \
+    --recipe zoo:nlp/masked_language_modeling/bert-base/pytorch/huggingface/wikipedia_bookcorpus/12layer_pruned80_quant-none-vnni?recipe_type=transfer-question_answering \
+```
+
+#### Knowledge Distillation
+By modifying the `distill_teacher` argument, you can enable [Knowledge Destillation](https://neptune.ai/blog/knowledge-distillation) (KD) functionality.
```
Comment (jeanniefinks) [inline src/sparseml/transformers/README_question_answering.md:62]:
```suggestion
By modifying the `distill_teacher` argument, you can enable [Knowledge Distillation](https://neptune.ai/blog/knowledge-distillation) (KD) functionality.
```

**Review by jeanniefinks — CHANGES_REQUESTED** [review]
Left some easy comments to consider.

File: src/sparseml/transformers/README_question_answering.md [hunk 16]
```diff
@@ -0,0 +1,151 @@
...
+  ...
+```
+
+To learn about the Hugging Face Transformers parameters in more detail, refer to [Hugging Face Transformers documentation](https://huggingface.co/docs/transformers/main_classes/trainer#transformers.TrainingArguments).
+
+## Once the Training is Done...
+
+The artifacts of the training process are saved to the directory `--output_dir`. Once the script terminates, you should find there everything required to deploy or further modify the model. 
+
+This includes the recipe (with the full description of the sparsification attributes), checkpoint files (saved in the appropriate framework format), Hugging Face Transformer's specific files (e.g. the tokenizer), etc.
```
[the remark this answers is not in the document]
Comment (dbogunowicz) [reply src/sparseml/transformers/README_question_answering.md:117]: @jeanniefinks Not sure whether I want to remove that information, sitting on the fence here. I do not like this paragraph very much, but I think it delivers an important message. Maybe rewording would be the best way out? Or maybe if a third reviewer also finds this paragraph unnecessary I would be ok with removing it.

[the remark this answers is not in the document]
Comment (bfineran) [reply src/sparseml/transformers/README_question_answering.md:117]: I would prefer to keep this change `Hugging Face Transformer's specific files (e.g. the tokenizer), etc.` to `configuration files for the tokenizer and dataset, etc.`

File: src/sparseml/transformers/README_question_answering.md [hunk 17]
```diff
@@ -0,0 +1,154 @@
...
+```
+
+### DeepSparse Engine Deployment
+
+Once the model is exported in the ONNX format, it is ready for deployment with the DeepSparse Engine. 
+
+The deployment is intuitive due to the DeepSparse Python API.
+
+```python
+from deepsparse.pipeline import Pipeline
```
Comment (bfineran) [inline src/sparseml/transformers/README_question_answering.md:137]:
```suggestion
from deepsparse import Pipeline
```

File: src/sparseml/transformers/README_question_answering.md [hunk 18]
```diff
@@ -0,0 +1,154 @@
...
+```
+
+To learn about the Hugging Face Transformers parameters in more detail, refer to [Hugging Face Transformers documentation](https://huggingface.co/docs/transformers/main_classes/trainer#transformers.TrainingArguments).
+
+## Once the Training is Done...
+
+The artifacts of the training process are saved to the directory `--output_dir`. Once the script terminates, the directory will have everything required to deploy or further modify the model such as:
+- The recipe (with the full description of the sparsification attributes).
+- Checkpoint files (saved in the appropriate framework format).
+- Hugging Face Transformer's specific files (e.g., the tokenizer).
```
Comment (bfineran) [inline src/sparseml/transformers/README_question_answering.md:115]:
```suggestion
- Additional configuration files (e.g. tokenizer, dataset info)
```

File: src/sparseml/transformers/README_question_answering.md (hunk 16 above)
[the remark this answers is not in the document]
Comment (bfineran) [reply src/sparseml/transformers/README_question_answering.md:117]: actually, after reading the context above, we can definitely remove as this repeats

**Review by bfineran — DISMISSED** [review]
LGTM

**Review by mgoin — DISMISSED** [review]
looks nice!

File: src/sparseml/transformers/README_question_answering.md [hunk 19]
```diff
@@ -0,0 +1,153 @@
...
+```
+
+To learn about the Hugging Face Transformers parameters in more detail, refer to [Hugging Face Transformers documentation](https://huggingface.co/docs/transformers/main_classes/trainer#transformers.TrainingArguments).
+
+## Once the Training is Done...
+
+The artifacts of the training process are saved to the directory `--output_dir`. Once the script terminates, the directory will have everything required to deploy or further modify the model such as:
+- The recipe (with the full description of the sparsification attributes).
+- Checkpoint files (saved in the appropriate framework format).
+- Additional configuration files (e.g. tokenizer, dataset info)
```
Comment (jeanniefinks) [inline src/sparseml/transformers/README_question_answering.md:115]:
```suggestion
- Additional configuration files (e.g., tokenizer, dataset info).
```

**Review by jeanniefinks — APPROVED** [review]
@bfineran added the tiny suggested edit. Make it happen and I am pre-approving it now!

**Review by mgoin — APPROVED** [review]
ship it!

**Review by mgoin — APPROVED** [review]
